ipld/get.go

```diff
@@ -148,42 +150,102 @@ func GetSharesByNamespace(
 }
 
 // GetLeavesByNamespace returns all the leaves from the given root with the given namespace.ID.
-// If nothing is found it returns both data and err as nil.
+// If nothing is found it returns data as nil.
 func GetLeavesByNamespace(
 	ctx context.Context,
 	bGetter blockservice.BlockGetter,
 	root cid.Cid,
 	nID namespace.ID,
-) ([]ipld.Node, error) {
-	err := SanityCheckNID(nID)
-	if err != nil {
-		return nil, err
+) []ipld.Node {
+	type job struct {
+		id  cid.Cid
+		pos int
 	}
-	rootH := plugin.NamespacedSha256FromCID(root)
-	if nID.Less(nmt.MinNamespace(rootH, nID.Size())) || !nID.LessOrEqual(nmt.MaxNamespace(rootH, nID.Size())) {
-		return nil, nil
-	}
-	// request the node
-	nd, err := plugin.GetNode(ctx, bGetter, root)
-	if err != nil {
-		return nil, err
+
+	// TODO: There has to be a more elegant solution
+	// bookkeeping is needed to be able to sort the leaves after the walk
+	type result struct {
+		node ipld.Node
+		pos  int
 	}
-	// check links
-	lnks := nd.Links()
-	if len(lnks) == 1 {
-		// if there is one link, then this is a leaf node, so just return it
-		return []ipld.Node{nd}, nil
+
+	// TODO: Should this be NumWorkersLimit?
+	// we don't know the amount of shares in the namespace, so we cannot preallocate properly
+	jobs := make(chan *job, NumWorkersLimit)
+	jobs <- &job{id: root}
+
+	// the wg counter cannot be preallocated either, it is incremented with each job
+	wg := sync.WaitGroup{}
+	wg.Add(1)
+
+	var leaves []result
+	mu := &sync.Mutex{}
+
+	// once the wait group is done, we can close the jobs channel to break the following loop
+	go func() {
+		wg.Wait()
+		close(jobs)
+	}()
+
+	for j := range jobs {
```

Comment:
Considering we don't know how many loops we have, we can do a `do/while` style for loop which is stopped from the inside. Chan canceling routine might not be needed

Reply:
Yeah, will wrap with context :)
Regarding the do/while, that was my first approach but the issue is that the logic deciding if the loop should break depends on the `WaitGroup` being finished. The waitgroup's counter isn't exported, so what it would probably look like is wrapping the wg in a new struct where we keep track of the counter ourselves. Should I try that approach?

Reply:
Yes, the described approach should work, and its cheaper than spawning another goroutine